prover/src/logup_gkr/mod.rs

```diff
@@ -114,45 +119,72 @@ impl<E: FieldElement> EvaluatedCircuit<E> {
-        }
+            unsafe { utils::uninit_vector(main_trace.main_segment().num_rows() * num_fractions) };
+        let num_cols = main_trace.main_segment().num_cols();
+        let num_oracles = evaluator.get_oracles().len();
+        let num_periodic_cols = periodic_values.num_columns();
+
+        batch_iter_mut!(
+            &mut input_layer_wires,
+            1024,
+            |batch: &mut [CircuitWire<E>], batch_offset: usize| {
+                let mut main_frame = EvaluationFrame::new(num_cols);
+
+                let mut query = vec![E::BaseField::ZERO; num_oracles];
+                let mut periodic_values_row = vec![E::BaseField::ZERO; num_periodic_cols];
+                let mut numerators = vec![E::ZERO; num_fractions];
+                let mut denominators = vec![E::ZERO; num_fractions];
+                let num_rows_to_process = batch.len() / numerators.len();
+                let row_offset = batch_offset / numerators.len();
+                for i in (0..main_trace.main_segment().num_rows())
+                    .skip(row_offset)
+                    .take(num_rows_to_process)
+                {
+                    let wires_from_trace_row = {
+                        main_trace.read_main_frame(i, &mut main_frame);
+                        periodic_values.fill_periodic_values_at(i, &mut periodic_values_row);
+                        evaluator.build_query(&main_frame, &mut query);
+
+                        evaluator.evaluate_query(
+                            &query,
+                            &periodic_values_row,
+                            log_up_randomness,
+                            &mut numerators,
+                            &mut denominators,
+                        );
+                        if 0 <= 1 {};
+                        let input_gates_values: Vec<CircuitWire<E>> = numerators
+                            .iter()
+                            .zip(denominators.iter())
+                            .map(|(numerator, denominator)| {
+                                CircuitWire::new(*numerator, *denominator)
+                            })
+                            .collect();
+                        input_gates_values
+                    };
+
+                    for (j, value) in wires_from_trace_row.iter().enumerate() {
+                        // SAFETY:
+                        // We have that `(i - row_offset) * num_fractions <= (num_rows - 1) * num_fractions`
+                        // by virtue of `i - row_offset <= num_rows - 1`.
+                        // Since `j <= num_fractions - 1`, we get that
+                        // `(i - row_offset) * num_fractions + j <= num_rows * num_fractions - 1`.
+                        unsafe {
+                            let batch_entry =
+                                batch.get_unchecked_mut((i - row_offset) * num_fractions + j);
+                            *batch_entry = *value;
+                        }
+                    }
```

Comment:
Would `.copy_from_slice()` not work better here? For example, something like:
```rust
batch[row_offset * num_fractions..].copy_from_slice(&wires_from_trace_row);
```

Reply:
Works! Only that it is `(i - row_offset)` instead of `row_offset`